draft-ietf-ppm-dap.md

```diff
@@ -1490,33 +1521,73 @@ if the report share should continue being prepared:
+  a `PrepareStep` in the `continued` state containing `prep_msg`.
+
+If the `round` in the Leader's request is equal to the Helper's current round
+(i.e., this is not the first time the Leader has sent this request), then the
+Helper responds with the current round's prepare messages. The Helper MUST
+verify that the contents of the `AggregationJob` are identical to the previous
+message (see {{aggregation-round-skew-recovery}}).
+
+If the Leader's `round` is behind or more than one round ahead of the Helper's
+current round, then the Helper MUST abort with an error of type `roundMismatch`.
+
+If successful, the Helper responds to the Leader with HTTP status 200 OK, media
 type `application/dap-aggregation-job` and a body consisting of an
 `AggregationJob` (see {{aggregation-helper-init}}) compiled from the stepped
-`PrepareStep`s. The order of the steps MUST match that of the `PrepareStep`
-values in the leader's `AggregationJob`. The helper then awaits the next
-message from the leader.
-
-[[OPEN ISSUE: consider relaxing this ordering constraint. See issue#217.]]
+`PrepareStep`s.
+
+The Leader validates the `AggregationJob` according to the criteria in
+{{aggregation-job-validation}}. If the message is valid, the Leader moves to the
+next round of continuation with the enclosed prepare steps. Otherwise, the
+Leader should abandon the aggregation job entirely.
+
+#### Recovering From Round Skew {#aggregation-round-skew-recovery}
+
+`AggregationJob` messages contain a `round` field, allowing Aggregators to
+ensure that their peer is on an expected round of the VDAF preparation
+algorithm. In particular, the intent is to allow recovery from a scenario where
+the Helper successfully advances from round `n` to `n+1`, but its
+`AggregationJob` response to the Leader gets dropped due to something like a
+transient network failure. The Leader could then resend the request to have the
+Helper advance to round `n+1` and the Helper should be able to retransmit the
+`AggregationJob` that was previously dropped. To make that kind of recovery
+possible, Aggregator implementations SHOULD checkpoint preparation state to
+durable storage such that Leaders can re-construct continuation requests and
+Helpers can re-construct continuation responses as needed.
+
+When implementing a round skew recovery strategy, it is important for Helpers to
+ensure that the Leader's `AggregationJob` message did not change when it was
+re-sent (i.e., the two messages must contain the same set of report IDs and
+prepare messages). This rules out attacks in which the Leader could violate the
+privacy of a report by re-running a VDAF preparation round with different
+prepare messages. One way a Helper could address this would be to store a digest
+of the Leader's request, indexed by aggregation job ID and round, and refuse to
+service a request for a given aggregation job round unless it matches the
+previously seen request (if any).
```

Comment:
[@cjpatton points out](https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/395/files#r1070186916):

> Whether there is an attack here is speculation. I'm also not certain the defense is sufficient.

Should we drop this requirement until we have a stronger motivation for it? Or should we err on the side of caution?

Reply:
IMO, we should keep some language suggesting that the helper check that the retransmitted message matches the original message. Less strongly, IMO, this should be a requirement (but perhaps this could be a suggestion instead, e.g. a `SHOULD` instead of a `MUST`).

My concern is less about an attack (I'm also not sure there is an attack here[1]), and more about consistency/coherence -- it would be _extremely_ confusing if some buggy Leader behavior caused them to retransmit a different message the second time around, and they got an answer as if they had sent the original message.

[1] Not caused by this PR, but it looks like neither the leader nor the helper checks that the request & response have the same reports in the same order during aggregation continuance. (The spec does dictate the helper's responses `MUST` be in the same order, but no one is required to check this.) Am I missing something or should this be fixed?

Reply:
I'm fine with the requirement for checking consistency. One question I have: What triggers the consistency check? Is it noticing that for a given agg job, the current agg request has the same round number as the current one?

I also agree with Brandon: Consistency/coherence is a stronger case for this requirement. I think SHOULD is appropriate; if we want to lift to MUST, then I think we had better be clearer about the actual mechanism.

That said, I think we should leave a note about the potential attack vector. The problem is the Leader's ability to "rewind" and the Helper and retry on different inputs. For instance, if the VDAF verification key changes, then the Leader gets to see different verifier shares produced by the Helper. This is potentially an avenue of attack, and I'm not sure it's ruled out by the consistency check. We'll have to wait for security analysis to get more clarity here.

I think our best bet is to leave this as an OPEN ISSUE. Something like:

> [[OPEN ISSUE: Allowing the Leader to "rewind" the state of an aggregation job may be an avenue of attack on privacy. For instance, if the VDAF verification key changes, the preparation shares in the Helper's response would change even if the consistency check is made. Security analysis is required.]]

> @branlwyd [1] Not caused by this PR, but it looks like neither the leader nor the helper checks that the request & response have the same reports in the same order during aggregation continuance. (The spec does dictate the helper's responses MUST be in the same order, but no one is required to check this.) Am I missing something or should this be fixed?

Hmm, I believe this is supposed to be enforced by the Helper. It should be fixed if not.

Reply:
Chris:
> What triggers the consistency check? Is it noticing that for a given agg job, the current agg request has the same round number as the current one?

Yes. We want to make sure that the leader isn't doing exactly what you describe, rewinding state and re-running an aggregation round with different broadcast prep messages.

I think I like the idea of making this a SHOULD so that implementations can defend against innocent bugs, and we'll leave an OPEN ISSUE about whether this needs to be upgraded to a MUST as an attack mitigation.

Bran:
> [1] Not caused by this PR, but it looks like neither the leader nor the helper checks that the request & response have the same reports in the same order during aggregation continuance. (The spec does dictate the helper's responses MUST be in the same order, but no one is required to check this.) Am I missing something or should this be fixed?

At [the bottom of "Helper Initialization"](https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/blob/71ccb43c4356228c70219bf87f6c1ee4abadf8c2/draft-ietf-ppm-dap.md?plain=1#L1288) and [the bottom of "Helper Continuation"](https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/blob/71ccb43c4356228c70219bf87f6c1ee4abadf8c2/draft-ietf-ppm-dap.md?plain=1#L1560), we have text that states the leader has to apply the criteria in [`{{aggregation-job-validation}}`](https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/blob/71ccb43c4356228c70219bf87f6c1ee4abadf8c2/draft-ietf-ppm-dap.md?plain=1#L1421). Admittedly it's confusing that this leader behavior appears in sections for the helper. The intent is that the text flows like this:

- describe leader request
- describe helper handling of request and how helper responds
- describe leader handling of helper's response

Which follows the order in which this happens. We could move those paragraphs into "Leader Initialization" and "Leader Continuation", respectively, but then the flow of text would be:

- describe leader request
- describe leader handling of helper's response
- describe helper handling of request and how helper responds

I'll give that a try and see how it reads.

The other thing is that currently, the leader is allowed to change the order of reports across aggregation rounds. I believe that's already the case in DAP-03 (I don't see anything in [DAP-03 leader continuation](https://datatracker.ietf.org/doc/html/draft-ietf-ppm-dap-03#section-4.4.2.1) governing the order of `AggregateContinueReq.prepare_steps`. If we want to add ordering requirements on the leader side, I think we should discuss that in a separate PR.

Reply:
I put in a SHOULD and a TODO as discussed, and wrote up #401 to track the security analysis we discussed. Additionally I remixed the text on how the leader validates `AggregationJob` messages and I think it reads better now.

Reply:
Thanks -- IMO it's easier to follow with the leader checks in the Leader Initialization/Continuation sections, I like this change.